src/rhsmlib/file_monitor.py

```diff
@@ -0,0 +1,273 @@
+        self.notifier = None
+
+    def loop(self, callback=None):
+        """
+        sets up watch manager, notifier, adds watches to watch manager, and starts loop
+        :param callback: callback method to be called at the end of each iteration of the loop
+        :return:
+        """
+        self.watch_manager = pyinotify.WatchManager()
+        self.notifier = pyinotify.Notifier(self.watch_manager, self.handle_event)
+        self.add_watches()
+
+        def inotify_callback():  # arg: notifier, required for pyinotify
+            return end_loop_cb(self, user_callback=callback)
+
+        while not inotify_callback():
+            self.notifier.process_events()
+            if self.notifier.check_events(500):  # keeps tests reasonably fast while still timing out
+                self.notifier.read_events()
+
+    def handle_event(self, event):
+        """
+        default process function for pyinotify notifier
+        :param event: pyinotify Event object, has path and mask of flags representing file modification
+        :return:
+        """
+        for dir_watch in self.dir_watches:
+            if dir_watch.paths_match(event.path) and dir_watch.file_modified(event.mask):
+                dir_watch.notify()
+
+    def add_watches(self):
+        """
+        adds watches to the watch manager
+        :return:
+        """
+        for dir_watch in self.dir_watches:
+            self.watch_manager.add_watch(dir_watch.path, dir_watch.mask, do_glob=dir_watch.is_glob)
+
+
+class DirectoryWatch(object):
+    """
+    Directory to be watched
+
+    Included in list to be passed into filesystem watcher object
+    Example usage:
+        directory_watch = DirectoryWatch("~/home/", [my_function1, my_function2])
+    """
+    def __init__(self, path, callbacks, is_glob=False):
+        """
+        :param path: path associated with directory to be watched
+        :param callbacks: list of methods called when directory is changed
+        :param is_glob: bool - if path provided is glob or not
+        """
+        self.IN_CREATE = 0x00000100
+        self.IN_DELETE = 0x00000200
+        self.IN_MODIFY = 0x00000002
+
+        self.path = path
+        self.timestamp = None
+        self.is_glob = is_glob
+        self.callbacks = callbacks
+        self.mask = self.IN_CREATE | self.IN_DELETE | self.IN_MODIFY
+
+    def notify(self):
+        """
+        calls all callbacks associated with dir watch
+        :return:
+        """
+        for cb in self.callbacks:
+            try:
+                cb()
+            except Exception as e:
+                log.exception(e)
+
+    def paths_match(self, event_path):
+        """
+        checks if event path matches any of the dir watch paths associated to it (in case dir watch path is glob)
+        :param event_path: path of the inotify event object
+        :return: bool - if paths match or not
+        """
+        if self.is_glob:
+            paths = glob.glob(self.path)
```

Comment:
This will return the matching paths from the filesystem from the current directory. I think that this should only check to see if the event_path matches the path of the directory watch using fnmatch.fnmatch.

Then the below is not necessary.

Reply:
Is what I am currently doing not equivalent? I agree that using a built-in function is probably better but I want to make sure I have the logic right